bin/plan

```diff
@@ -1,42 +1,40 @@
-  [ "${1}" = "green" ] && _fg='\033[0;32m'
-  shift
-  # shellcheck disable=SC2145
-  echo -e "${_fg}>>> ${@}\033[0m"
-}
+# shellcheck source=.common.sh
+source "$(dirname ${BASH_SOURCE[0]})/.common.sh"
 
-changed_dirs=$(git diff-tree --no-commit-id --name-only -r $master_base_sha $branch_head_sha | awk 'BEGIN{FS=OFS="/"}/\/resources\//{NF-=1; print $3}' | sort | uniq)
+_checkenv
 
-for _c in namespaces/*; do
-  cluster="$(basename ${_c})"
-  log green "found cluster ${cluster}"
-  ( set -x; kubectl config use-context "${cluster}" ) || { log red "no context found, skipping ${cluster}" && continue; }
-  if [ -n "$changed_dirs" ]; then
-    for _f in ${changed_dirs}; do
-      namespace="$(basename ${_f})"
-      terraform_resources_path="namespaces/${cluster}/${namespace}/resources"
-      if [ -d "${terraform_resources_path}" ]; then
-        cd "${terraform_resources_path}"
-        log blue "Planning terraform resources for namespace ${namespace} in ${cluster}"
-        terraform init \
-          -backend-config="bucket=moj-cp-k8s-investigation-environments-terraform" \
-          -backend-config="key=${cluster}/${namespace}/terraform.tfstate" \
-          -backend-config="region=eu-west-1"
-        terraform plan \
-          -var="cluster_name=${cluster%.k8s.integration.dsd.io}" \
-          -var="cluster_state_bucket=moj-cp-k8s-investigation-platform-terraform"
-        cd $OLDPWD
-      fi
-    done
-  else
-    log blue "No terraform changes in this PR"
-  fi
-done
\ No newline at end of file
+cluster="${PIPELINE_CLUSTER}"
+
+# The two undefined variables below (master_base_sha and branch_head_sha)
+# are passed by the github-pull-request concourse resource.
+# shellcheck disable=SC2154
+changed_dirs=$(git diff-tree --no-commit-id --name-only -r $master_base_sha $branch_head_sha | awk 'BEGIN{FS=OFS="/"}/\/resources\//{NF-=1; print $3}' | sort | uniq | grep "${PIPELINE_CLUSTER}") || true
+
+log green "planning for cluster ${cluster}"
+( set -x; kubectl config use-context "${cluster}" ) || { log red "no context found, skipping ${cluster}" && exit 1; }
+if [ -n "$changed_dirs" ]; then
+  for _f in ${changed_dirs}; do
+    namespace="$(basename ${_f})"
+    terraform_resources_path="namespaces/${cluster}/${namespace}/resources"
+    if [ -d "${terraform_resources_path}" ]; then
+      cd "${terraform_resources_path}"
+      log blue "Planning terraform resources for namespace ${namespace} in ${cluster}"
+      terraform init \
+        -backend-config="bucket=${PIPELINE_STATE_BUCKET}" \
+        -backend-config="key=${PIPELINE_STATE_KEY_PREFIX}${cluster}/${namespace}/terraform.tfstate" \
+        -backend-config="region=${PIPELINE_STATE_REGION}"
+      terraform plan \
+        -var="cluster_name=${cluster%%.*}" \
+        -var="cluster_state_bucket=${PIPELINE_CLUSTER_STATE_BUCKET}" \
+        -var="cluster_state_key=${PIPELINE_CLUSTER_STATE_KEY_PREFIX}${cluster%%.*}/terraform.tfstate"
+      cd $OLDPWD
```

Comment:
Where is OLDPWD being set?

Also, you can avoid needing to do that kind of thing by wrapping your stuff in () to invoke a sub-shell, e.g.

```
(
    cd foo
    do some stuff
)
```

After this, the script will continue with pwd being whatever it was before the (

Reply:
`OLDPWD` is a shell variable set by `cd` - I'm not sure whether it's an official POSIX shell variable (although it seems to be implemented) but is definitely part of bash https://www.tldp.org/LDP/abs/html/internalvariables.html

Using a sub-shell is much nicer (and we actually use it above!) indeed, will make the adjustment .